### PR TITLE
Ticker

### PR DESCRIPTION
**Pull request by wilsonjwco**
Feature add. Added a scrolling ticker on the digital watch face. Ticker has 3 modes controlled by a settings screen.

1. Off: Do not display the ticker.
2. Update: Display the most recent message. As a convenience, the last message can be seen at a glance without the need to swipe up.
3. Keep: Keep the currently displayed message, and do not update it when receiving new messages. Useful for persistent reminders or to help with learning/memorizing facts, Bible verses, quotes, etc. Message can be set manually using GadgetBridge's Debug screen.

Compiled, downloaded OTA to my watch, and tested. 

Pics can be seen on this google shared album: [https://photos.app.goo.gl/HSDs2YWnEf5c1Xk48](https://photos.app.goo.gl/HSDs2YWnEf5c1Xk48)

Ideas for future functionality:

- display news/RSS items
- stock market ticker
- user-programmable filter to show only high priority or certain types of messages on the watch face
- display status from other apps (notify when timer is done, or when step goal met)
- debug/log messages for developers trying to catch rare or hard to repeat bugs

### CODE REVIEW

File: src/components/settings/Settings.h [hunk 1]
```diff
@@ -77,6 +78,15 @@ namespace Pinetime {
         return settings.wakeUpMode;
       };
 
+      void setTickerMode(TickerMode ticker) {
+        if (ticker != settings.tickerMode)
```
Comment (Avamander) [inline src/components/settings/Settings.h:82]: Missing braces

File: src/displayapp/screens/WatchFaceDigital.cpp [hunk 2]
```diff
@@ -100,6 +101,33 @@ WatchFaceDigital::WatchFaceDigital(DisplayApp* app,
...
+
+  /*
+  if (title == nullptr)
+    title = "Notification";
+  char* pchar;
+  pchar = strchr(title, '\n');
+  while (pchar != nullptr) {
+    *pchar = ' ';
+    pchar = strchr(pchar + 1, '\n');
+  }
```
Comment (Avamander) [inline src/displayapp/screens/WatchFaceDigital.cpp:113]: Unnecessary comment

File: src/displayapp/screens/WatchFaceDigital.cpp [hunk 3]
```diff
@@ -129,12 +157,13 @@ bool WatchFaceDigital::Refresh() {
 
   notificationState = notificatioManager.AreNewNotificationsAvailable();
   if (notificationState.IsUpdated()) {
-    if (notificationState.Get() == true)
+    if (notificationState.Get() == true) {
       lv_label_set_text(notificationIcon, NotificationIcon::GetIcon(true));
-    else
+    } else
```
Comment (Avamander) [inline src/displayapp/screens/WatchFaceDigital.cpp:162]: Missing brace

File: src/displayapp/screens/WatchFaceDigital.h [hunk 4]
```diff
@@ -57,6 +60,7 @@ namespace Pinetime {
         lv_obj_t* label_time;
         lv_obj_t* label_time_ampm;
         lv_obj_t* label_date;
+	lv_obj_t* label_ticker;
```
Comment (Avamander) [inline src/displayapp/screens/WatchFaceDigital.h:63]: Run clang-format on these files.

File: src/displayapp/screens/settings/SettingTicker.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        if (i == 0) {
+          settingsController.setTickerMode(Pinetime::Controllers::Settings::TickerMode::Off);
+        };
+        if (i == 1) {
+          settingsController.setTickerMode(Pinetime::Controllers::Settings::TickerMode::Update);
+        };
+        if (i == 2) {
+          settingsController.setTickerMode(Pinetime::Controllers::Settings::TickerMode::Keep);
+        };
```
Comment (Avamander) [inline src/displayapp/screens/settings/SettingTicker.cpp:95]: Unnecessary semicolons

**Review by Avamander — CHANGES_REQUESTED** [review]

**Comment by Riksu9000** [comment]
Maybe instead of having modes, we could simply pin a notification to show on the watchface.

**Comment by 0x1a8510f2** [comment]
Just so you're aware @wilsonjwco, you haven't removed geolocation data from the photos in the album, meaning that anyone can see where the photo was taken (likely where you live). Would recommend you delete the album and re-create it without geolocation data (there should be a setting in Google photos).

**Comment by wilsonjwco** [comment]
> Maybe instead of having modes, we could simply pin a notification to show on the watchface.

Do you mean we could have a button on the watch face that would pin the current notification?

**Comment by wilsonjwco** [comment]
Thanks everyone for the comments, I'll work through them. Also noticed the message get corrupted after a few wake cycles, so it looks like I've got more debugging to do before it's merged.

**Comment by Riksu9000** [comment]
> > Maybe instead of having modes, we could simply pin a notification to show on the watchface.
> 
> @Riksu9000 do you mean we could have a button on the watch face that would pin the current notification?

No. I was thinking something like opening the notification that we want to pin and long pressing on it. Then when we return to the watch face, it would be shown there. This way we could easily choose any notification to keep.

**Comment by wilsonjwco** [comment]
> I was thinking something like opening the notification that we want to pin and long pressing on it.

Great idea-- I'll get rid of the settings and make it like you suggested. Will follow up with this discussion once I've done this, fixed the above mentioned problems, and am ready for the merge to be considered again.

**Comment by lman0** [comment]
any new on this @wilsonjwco ?

**Comment by Riksu9000** [comment]
I'll close this since it's very outdated. Please create a new PR if you want to keep working on the notification pinning feature.

**Pull request closed without merging**